### DISCUSSION TITLE
emit the event on behalf of the client 

### DISCUSSION THREAD

**Discussion by hemanth-reckonsys** [miguelgrinberg/python-socketio#1105] [Q&A]
Hello, I'm using a socket. io WSGIApp, eventlet, and RedisManager (public Redis url) as the first server (S1 is running fine). Now that I have another Django server (separate isolated server s2), I need to emit a disconnect event on behalf of the client to my first server (s1) from s2, using the sid that I saved in the database,i  blocked that sid and I need to disconnect. How do you do that?

**Comment by miguelgrinberg** [comment]
I'm not sure I understand what you want to do.

A server cannot emit on behalf of a client. Your s1 server can directly emit to clients. Your s2 server (assuming you configure it as an auxiliary process) can also emit to clients, indirectly through s1.

Another confusing aspect of your description is that you want to emit a disconnect event, but Socket.IO has no such thing. If you want to remove a client, then the server can disconnect it, but this isn't done with the emit function, there is a disconnect function for this.

**Comment by hemanth-reckonsys** [comment]
Hi, usually the client's(front-end) connection is idle(no response from frontend) for some time(idle more than 15 minutes(because a long idle connection is consuming RAM))), so I want to emit a disconnect event to close the idle connection from s2 to s1.
for example
```
from socketio import RedisManager, Server
sio = Server(client_manager=mgr)
sio.disconnent(sid)
```

**Reply by miguelgrinberg** [reply]
The `Server` object is used only for servers, so s1 in your case. In s2, you have to use the `RedisManager` object:

```python
mgr = RedisManager('redis://', write_only=True)
mgr.disconnect(sid)
```

Documentation: https://python-socketio.readthedocs.io/en/latest/server.html#emitting-from-external-processes

**Reply by hemanth-reckonsys** [reply]
`TypeError: disconnect() missing 1 required positional argument: 'namespace'`

**Reply by miguelgrinberg** [reply]
Here is the documentation for the disconnect function: https://python-socketio.readthedocs.io/en/latest/api.html#socketio.BaseManager.disconnect. Add the namespace and you'll be fine. If you are using the default namespace, then use `'/'`.

**Comment by hemanth-reckonsys** [comment]
ok,
my s1 
```
mgr = RedisManager("redis://globalxxxxxxawsxxxxxx",)                              
sio = Server( 
   ...:     client_manager=mgr, 
   ...:     cors_allowed_origins="*", 
   ...:     ping_interval=25, 
   ...:     ping_timeout=120, 
   ...: )   

```
 i tried from django shell(s2)

```
mgr = RedisManager("redis://globalxxxxxxawsxxxxxx", write_only=True)                              

In [5]: sio = Server( 
   ...:     client_manager=mgr, 
   ...:     cors_allowed_origins="*", 
   ...:     ping_interval=25, 
   ...:     ping_timeout=120, 
   ...: )                                                                                                                           
INFO 2022-12-28 16:38:40,376 server 82040 140347691530045 Server initialized for eventlet.

In [6]: sio.disconnect("756b6ebb1fb346eba435218920b7bfdc")
ERROR 2022-12-28 16:41:42,885 redis_manager 82040 140347691530045 Cannot publish to redis... retrying
ERROR 2022-12-28 16:43:53,957 redis_manager 82040 140347691530045 Cannot publish to redis... giving up
```
Because s2 did not close the s1 connection, the socket connection is still not disconnected.

**Reply by miguelgrinberg** [reply]
This is wrong. As I said above, s2 is not a Socket.IO server, so don't use the `Server` class.

**Reply by hemanth-reckonsys** [reply]
ok, 
s1
```
mgr = RedisManager("redis://globalxxxxx",)                              
sio = Server( 
   ...:     client_manager=mgr, 
   ...:     cors_allowed_origins="*", 
   ...:     ping_interval=25, 
   ...:     ping_timeout=120, 
   ...: )

@sio.event
def disconnect(sid):
    print("disconnect", sid)

eventlet.monkey_patch()
app = WSGIApp(sio)
eventlet.wsgi.server(eventlet.listen(("", 8000)), app)
```
s2
```
from socketio import RedisManager, Server                                                                                   
In [2]: mgr = RedisManager("redis://globalxxxxx", write_only=True)                              
In [3]: mgr.disconnect("5cba6eacb22541efb9a9f82fe03d8a6e", "/")
```
no luck , s1 disconnect(I added a print statement inside the function to verify) function is not called, if I close the browser tab or the client emits a disconnect event then only the disconnect print function executed.

**Reply by miguelgrinberg** [reply]
I'm having difficulty understanding what you are doing. First of all, S1 and S2 have to connect to the same Redis service so that they can communicate. You have different URLs. Beyond that, please see the troubleshooting section of the docs to learn how to add logging to your server, then show me the logs.

**Reply by hemanth-reckonsys** [reply]
i added my complete configuration
in simple s1 to f1 (working fine) I need to kill the established socket connection between s1 and f1 from s2 using the sid(on behalf of the client f1)

**Reply by miguelgrinberg** [reply]
The client can disconnect itself, it does not need to ask a server to do it on its behalf.

**Reply by hemanth-reckonsys** [reply]
Yes, the connection is idle in my case, so I have to invoke disconnect on that specific SID otherwise the socket will not release the RAM and is this can lead to a memory leak?

**Reply by miguelgrinberg** [reply]
That's fine, but then you are not disconnecting "on behalf of the client". You are effectively removing a client. You did not do what I asked though:

> please see the troubleshooting section of the docs to learn how to add logging to your server, then show me the logs.
